### PR TITLE
Convenience features for .SDcols patterns perl=TRUE

### LINKED ISSUE #5387: FR: Allow `patterns` to have the arguments of `grep`
Beyond the 2 essential arguments of `grep`, most of the other would be also useful if they were included in `patterns`. In fact, any other except `value` could help to look for much other patterns. 

In particular, it is needed for certain #regex which only work under `perl=TRUE`.

So I propose to modify
```r
patterns = function(..., cols=character(0L)) {
  # if ... has no names, names(list(...)) will be "";
  #   this assures they'll be NULL instead
  L = list(...)
  p = unlist(L, use.names = any(nzchar(names(L))))
  if (!is.character(p))
    stopf("Input patterns must be of type character.")
  matched = lapply(p, grep, cols)
  # replace with lengths when R 3.2.0 dependency arrives
  if (length(idx <- which(sapply(matched, length) == 0L)))
    stopf('Pattern(s) not found: [%s]', brackify(p[idx]))
  matched
}
```
with something like
```r
patterns = function(..., cols=character(0L), ignore.case = FALSE, perl = FALSE, fixed = FALSE, useBytes = FALSE, invert = FALSE) {
  # if ... has no names, names(list(...)) will be "";
  #   this assures they'll be NULL instead
  L = list(...)
  p = unlist(L, use.names = any(nzchar(names(L))))
  if (!is.character(p))
    stopf("Input patterns must be of type character.")
  matched = lapply(p, grep, cols, ignore.case = ignore.case, perl = perl, fixed = fixed, useBytes = useBytes, invert = invert)
  # replace with lengths when R 3.2.0 dependency arrives
  if (length(idx <- which(sapply(matched, length) == 0L)))
    stopf('Pattern(s) not found: [%s]', brackify(p[idx]))
  matched
}
```

Thanks!

### PR DESCRIPTION
**Pull request by tdhock**
Closes #5387

### CODE REVIEW

**Comment by tdhock** [comment]
hi @iago-pssjd I implemented a PR which hopefully resolves your issue, can you please review?
```r
> DT = data.table(a=1:2, b=3, c='a')
> DT[, .SD, .SDcols=patterns('^(?!W)a',perl=TRUE)]
       a
   <int>
1:     1
2:     2
> not = function(regex, cols)grep(regex, cols, invert=TRUE)
> DT[, .SD, .SDcols=not("[ac]")]
       b
   <num>
1:     3
2:     3
```

**Comment by iago-pssjd** [comment]
Hi @tdhock! Thanks for your work, I tested your branch and it works fine regarding the issue.

**Comment by tdhock** [comment]
hi @jangorecki @MichaelChirico I would like to include this fix in next release, and I believe that it is ready to merge into master, so could you please review and/or merge if you agree that it is ready? Thanks!

**Comment by MichaelChirico** [comment]
Let's sort out #5937 first

File: NEWS.md
Comment (MichaelChirico) [inline NEWS.md:1]: update the NEWS entry?

File: man/data.table.Rd [hunk 1]
```diff
@@ -145,18 +145,18 @@ data.table(\dots, keep.rownames=FALSE, check.names=FALSE, key=NULL, stringsAsFac
...
+    Finally, you can filter columns to include in \code{.SD} based on their \emph{names} according to regular expressions via \code{.SDcols=patterns(regex1, regex2, ...)}. The included columns will be the \emph{intersection} of the columns identified by each pattern; pattern unions can easily be specified with \code{|} in a regex. User-defined functions, \code{.SDcols=user_fun()} with an argument named \code{cols} will get data table column names, and should return a vector or list of vectors (combined via intersection). You can filter columns on \code{values} by passing a function, e.g. \code{.SDcols=\link{is.numeric}}. You can also invert a pattern as usual with \code{.SDcols=!patterns(...)} or \code{.SDcols=!is.numeric}.
```
Comment (MichaelChirico) [inline man/data.table.Rd:154]: let's omit mention of the `user_fun()` part here, and figure out how to proceed there in #5115 instead?

File: man/data.table.Rd
Comment (MichaelChirico) [inline man/data.table.Rd:1]: is there any substantive change to this file related to `perl=TRUE` addition? It looks like it's just formatting changes.

The formatting changes are an improvement, but a broader PR fixing the issue would be good too (other .Rd files are similarly affected):

https://github.com/search?q=repo%3ARdatatable%2Fdata.table+path%3Aman+%2F%5Cn%5B%5E%23%5Cn%5D*%60.*%60%2F&type=code

File: R/fmelt.R [hunk 2]
```diff
@@ -19,14 +19,14 @@ melt.default = function(data, ..., na.rm = FALSE, value.name = "value") {
   # nocov end
 }
 
-patterns = function(..., cols=character(0L)) {
+patterns = function(..., cols=character(0L), perl=FALSE) {
```
Comment (MichaelChirico) [inline R/fmelt.R:22]: why not include the rest of the `grep()` signature while we're at it, namely `ignore.case`, `fixed`, `useBytes` could be useful. `value` and `invert` don't seem so useful.

File: man/patterns.Rd [hunk 3]
```diff
@@ -12,14 +12,15 @@ and melt them into separate columns. See the \code{Efficient reshaping using
...
+patterns(\dots, cols=character(0), perl=FALSE)
 }
 \arguments{
   \item{\dots}{A set of regular expression patterns.}
   \item{cols}{A character vector of names to which each pattern is matched.}
+  \item{perl}{Passed to \code{grep}: \code{TRUE} to use PCRE, \code{FALSE} to use TRE.}
 }
 \seealso{
-  \code{\link{melt}},
+  \code{\link{melt}}, \code{\link{grep}}
```
Comment (MichaelChirico) [inline man/patterns.Rd:23]: somewhat unusual -- why not use `\code{\link{grep}}` in the `\item{perl}` entry above? That is where I would expect to find the link as a reader.

File: man/patterns.Rd [hunk 4]
```diff
@@ -12,14 +12,15 @@ and melt them into separate columns. See the \code{Efficient reshaping using
 data.tables} vignette linked below to learn more.
 }
 \usage{
-patterns(\dots, cols=character(0))
+patterns(\dots, cols=character(0), perl=FALSE)
 }
 \arguments{
   \item{\dots}{A set of regular expression patterns.}
   \item{cols}{A character vector of names to which each pattern is matched.}
+  \item{perl}{Passed to \code{grep}: \code{TRUE} to use PCRE, \code{FALSE} to use TRE.}
```
Comment (MichaelChirico) [inline man/patterns.Rd:20]:
```suggestion
  \item{perl}{Passed to \code{\link{grep}}.}
```

Probably better to just point to `?grep`; R users that don't know what `perl=` does yet are unlikely to understand "PCRE" or "TRE" either, I reckon.

File: R/data.table.R [hunk 5]
```diff
@@ -1019,9 +1019,12 @@ replace_dot_alias = function(e) {
             # .SDcols is of the format a:b, ensure none of : arguments is a call data.table(V1=-1L, V2=-2L, V3=-3L)[,.SD,.SDcols=-V2:-V1] #4231
             .SDcols = eval(colsub, setattr(as.list(seq_along(x)), 'names', names_x), parent.frame())
           } else {
-            if (colsub %iscall% 'patterns') {
-              # each pattern gives a new filter condition, intersect the end result
-              .SDcols = Reduce(intersect, eval_with_cols(colsub, names_x))
+            if (is.call(colsub) && !is.null(list_or_vector <- eval_with_cols(colsub, names_x))) {
```
Comment (MichaelChirico) [inline R/data.table.R:1022]: is this change related to the current PR?

File: inst/tests/tests.Rraw [hunk 6]
```diff
@@ -16611,6 +16611,14 @@ test(2128.2, names(DT[, .SD, .SDcols=!is.numeric]), 'c')
...
 test(2128.4, DT[, .SD, .SDcols=function(x) 2L],   error='conditions were not met for: [a, b, c]')
 test(2128.5, DT[, .SD, .SDcols=function(x) NA],   error='conditions were not met for: [a, b, c]')
+# user-defined pattern matching functions, PR#5663
+patterns = function(cols)grep('^(?!W)a', cols, perl=TRUE)
+test(2128.6, names(DT[, .SD, .SDcols=patterns()]), 'a')
+rm(patterns)
+not = function(regex, cols)grep(regex, cols, invert=TRUE)
+test(2128.7, names(DT[, .SD, .SDcols=not("[bc]")]), 'a')
+# patterns with PCRE.
+test(2128.8, names(DT[, .SD, .SDcols=patterns('^(?!W)a',perl=TRUE)]), 'a')
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:16621]: WDYT about making the test such that the perl pattern actually matters? e.g. a column named `.a` should be excluded

File: man/patterns.Rd (hunk 3 above)
Comment (tdhock) [reply man/patterns.Rd:23]: ok, I removed it

File: R/fmelt.R (hunk 2 above)
Comment (tdhock) [reply R/fmelt.R:22]: ok I added them

File: man/data.table.Rd (hunk 1 above)
Comment (tdhock) [reply man/data.table.Rd:154]: ok I reverted it.

File: NEWS.md
Comment (tdhock) [reply NEWS.md:1]: ok

File: man/data.table.Rd
Comment (tdhock) [reply man/data.table.Rd:1]: yes these are random unrelated formatting improvements, sorry for not separating it into a different PR

File: R/data.table.R (hunk 5 above)
Comment (tdhock) [reply R/data.table.R]: it was required to support user-defined functions but I reverted the change

File: inst/tests/tests.Rraw (hunk 6 above)
Comment (tdhock) [reply inst/tests/tests.Rraw]: I revised the pattern so that the lookahead is used

**Comment by tdhock** [comment]
ok I simplified the PR and addressed most of your comments, sorry that was so complicated initially

File: man/patterns.Rd [hunk 7]
```diff
@@ -12,11 +12,18 @@ and melt them into separate columns. See the \code{Efficient reshaping using
...
-patterns(\dots, cols=character(0))
+patterns(
+  \dots, cols=character(0),
+  ignore.case=FALSE, perl=FALSE,
+  fixed=FALSE, useBytes=FALSE)
 }
 \arguments{
   \item{\dots}{A set of regular expression patterns.}
   \item{cols}{A character vector of names to which each pattern is matched.}
+  \item{ignore.case}{Passed to \code{\link{grep}}.}
```
Comment (MichaelChirico) [inline man/patterns.Rd:23]: I think Rd supports like \item{ignore.case,perl,fixed,useBytes}{Passed...} which may be more appropriate here

**Comment by MichaelChirico** [comment]
> sorry that was so complicated initially

no problem! it's totally fine. all of the improvements are good & should happen, i'm just trying to instill a lot of discipline on keeping PRs focused. One thing to consider is a stacked PR flow (e.g. I've been using Graphite for this) -- when a PR starts getting more complicated, break it into a chain/stack of easily-digested PRs.

File: man/patterns.Rd [hunk 8]
```diff
@@ -12,11 +12,18 @@ and melt them into separate columns. See the \code{Efficient reshaping using
...
+  \item{perl}{Passed to \code{\link{grep}}.}
+  \item{fixed}{Passed to \code{\link{grep}}.}
+  \item{useBytes}{Passed to \code{\link{grep}}.}
```
Comment (MichaelChirico) [inline man/patterns.Rd:26]:
```suggestion
  \item{ignore.case, perl, fixed, useBytes}{Passed to \code{\link{grep}}.}
```

File: NEWS.md [hunk 9]
```diff
@@ -16,6 +16,8 @@
 
 2. `cedta()` now returns `FALSE` if `.datatable.aware = FALSE` is set in the calling environment, [#5654](https://github.com/Rdatatable/data.table/issues/5654).
 
+3. `data.table:::patterns` now accepts arguments `ignore.case`, `perl`, `fixed`, and `useBytes`, which are passed to `grep`, #5387. Thanks to @iago-pssjd for the feature request, and @tdhock for the implementation.
```
Comment (MichaelChirico) [inline NEWS.md:19]:
```suggestion
3. `patterns()` helper for `.SDcols` now accepts arguments `ignore.case`, `perl`, `fixed`, and `useBytes`, which are passed to `grep`, #5387. Thanks to @iago-pssjd for the feature request, and @tdhock for the implementation.
```

**Review by MichaelChirico — APPROVED** [review]
